SproxydKeysScan/Logging.js

```diff
@@ -0,0 +1,110 @@
+const werelogs = require('werelogs');
+const { env } = require('./env');
+
+/**
+ * @class
+ * @classdesc - aggregates events in a map and logs them at an interval.
+ * call .run() after instantiation to being loop.
+ */
+class AggregateLogger {
+    /**
+     * @constructor
+     * @param {number} logInterval - log summary statistics once every intervalSize seconds
+     */
+    constructor(logInterval) {
+        this.log = new werelogs.Logger('s3utils:SproxydKeysScan:AggregateLogger');
+        if (AggregateLogger._instance) {
+            const message = 'AggregateLogger is a singleton and has been instatiated. This instantiation is ignored';
+            AggregateLogger._instance.log.debug(message);
+            return AggregateLogger._instance;
+        }
+        this.interval = logInterval;
+        this.beginTime = new Date();
+        this.statistics = new Map();
+        AggregateLogger._instance = this;
+        return AggregateLogger._instance;
+    }
+
+    /**
+     * @param {string} event - increases frequency of event string by 1 in this.statistics.
+     * @returns {undefined}
+     */
+    update(event) {
+        this.statistics.set(event, (this.statistics.get(event) || 0) + 1);
+    }
+
+    /**
+     * logs the current event frequencies in this.statistics and resets the counts for the next interval.
+     * @returns {undefined}
+     */
+    logInterval() {
+        const currentTime = new Date();
+        const summary = {};
+        for (const [k, v] of this.statistics) {
+            summary[k] = v;
+        }
+        const data = {
+            beginTime: this.beginTime.toISOString(),
+            endTime: currentTime.toISOString(),
+            summary,
+        };
+        this.log.info('Summary update:', data);
+
+        // setup for next interval
+        this.beginTime = currentTime;
+        this.statistics.clear();
+    }
+
+    run() {
+        setInterval(this.logInterval.bind(this), this.interval * 1000);
+    }
+}
+/**
+ * @class
+ * @classdesc - creates a proxy to a werelogs.Logger class instance.
+ */
+class ProxyLoggerCreator {
+    /**
+    * @constructor
+    * @param {class} logger - an instance of werelogs.Logger
+    * @returns {class} - a proxy to the logger
+    */
+    constructor(logger) {
+        this.logger = logger;
+        this.aggregateLogger = new AggregateLogger(env.LOG_INTERVAL);
+        const context = this;
+        const logLevels = new Set(['trace', 'debug', 'info', 'warn', 'error', 'fatal']);
+        const handlers = {
```

Comment:
This whole proxy and reflect seems a bit complex for wrapping the werelogs logger, also seems a bit fragile with the "is function" check and log level check.

Reply:
I originally was thinking of just subclassing werelogs.Logger but some concerns were raised so I used a proxy pattern to not touch the underlying class. The checks are just saying "only intercept function calls with the names in logLevels". From what I've seen, its pretty standard usage.

Reply:
Personally I have never seen or used Proxy or Reflect here, I would just manually create all the methods but I guess its fine. Seems a little strange to pass in a logger to just override its methods instead of creating one but thats fine I guess.